### ISSUE TITLE
[Question] GLM-5.3-Flash on 4x GB200: `--moe-runner-backend flashinfer_trtllm` scores ~1.3 gsm8k points below `deep_gemm` in repeated runs (6 vs 3 runs); one `triton` run agrees with deep_gemm

### ISSUE THREAD

**Issue by NolenLiang** [sgl-project/sglang#39797]
### Describe the bug

In the TEP4 configuration we tested (TP4 + EP4, trtllm DSA backends, fp8 KV cache, `--moe-runner-backend flashinfer_trtllm` — the cookbook's GB200 cell is TP4 without EP and gave the same score, see below), the mean of six GLM-5.3-Flash 5-shot greedy gsm8k runs is ~1.3 points below the mean of three runs of the same server with `--moe-runner-backend deep_gemm`, and every one of the six trtllm-runner scores is below every one of the three deep_gemm scores (closest pair 0.9272 vs 0.9280); it is also below vLLM on the same hardware. A single `triton` run landed with deep_gemm. lm-evaluation-harness 0.4.13, `local-completions`, one prompt per request, all 1319 questions, flexible-extract.

Same server command repeated (one server start per run, 32 concurrent requests; sample sd):

| configuration | runs | mean | sample sd | values |
|---|---|---|---|---|
| SGLang TEP4 `--moe-runner-backend flashinfer_trtllm` | 6 | 0.9188 | 0.0048 | 0.9174, 0.9143, 0.9143, 0.9189, 0.9272, 0.9204 |
| SGLang TEP4 `--moe-runner-backend deep_gemm` | 3 | 0.9315 | 0.0032 | 0.9325, 0.9280, 0.9340 |
| SGLang TEP4 `--moe-runner-backend triton` | 1 | 0.9303 | | |
| vLLM TP4, FlashInfer trtllm FP8 MoE backend (same kernel) | 5 | 0.9327 | 0.0031 | 0.9340, 0.9272, 0.9348, 0.9340, 0.9333 |

Single trtllm-runner runs under changed conditions stay in the same range: TP4 without EP 0.9158; `--kv-cache-dtype bfloat16` 0.9151; strictly sequential requests 0.9174 / 0.9174; the exact cookbook cell on the cookbook image (`lmsysorg/sglang:glm-5.3-flash`, `0.0.0.dev1+gf609d677b`) 0.9181; a patch casting routing logits/bias to fp32 before the kernel (they already are fp32) 0.9265 / 0.9174. Not controls: "auto" DSA backends resolve to trtllm anyway (0.9181), and `--disable-shared-experts-fusion` is already set automatically for the trtllm runner (0.9212). Paired exact McNemar tests of the three original trtllm runs vs the first deep_gemm run: p = 0.020 / 0.0049 / 0.0043. The discordant questions are "different numeric answer", not extraction failures.

### What we checked (all negative)

- Same FlashInfer 0.6.18 `trtllm_fp8_block_scale_moe` kernel as vLLM's backend. Logged at run time (prefill and decode calls): fp32 router logits, fp32 correction bias, fp8 activations with a contiguous fp32 [H/128, T] scale tensor, n_group 1 / topk_group 1 / top_k 8 / routed_scaling_factor 2.5 / DeepSeekV3 routing, 72 local of 288 experts, `use_shuffled_weight=False` (`WeightLayout.MajorK`).
- Kernel-level check on real layer weights (layers 10 and 20; FlashInfer's own tests/moe reference helpers): with fp32 routing inputs the kernel is within ~4.5% relative L2 of an fp32 evaluation of the same FP8 tensors; unshuffled vs shuffled layouts bitwise identical; same for 1–4096 tokens, autotune on/off, in-kernel vs pre-routed top-k. Feeding the correction bias in bf16 would raise the error to 54% (the bias is 6.18–6.56 with a 0.38 spread), but SGLang passes fp32.
- Dumps of the exact tensors the runner hands to the kernel (start-up forwards): weights and block scales bitwise equal to the checkpoint (kernel [up; gate] row order), activation quantisation correct, fp32 logits and bias on prefill- and decode-sized calls. On the first two real lm-eval prompts (911 / 952 tokens, the same prompts for both runners; prefill only — decode runs as CUDA-graph replay and cannot be hooked from Python) the kernel output vs an fp32 evaluation of the checkpoint experts on each runner's own layer inputs: trtllm median 5.0% relative L2 (max 5.6%), deep_gemm median 5.3% (max 6.1%; it runs on UE8M0-requantised weights). Same reference definition, same prompts, but each runner sees its own layer inputs and routing, so not a paired comparison; the local per-layer errors are of the same order and do not explain the end-to-end gap.

### Question

Switching the runner reproducibly moves the score by ~1.3 points; we have not isolated the mechanism. What else differs between the `flashinfer_trtllm` and `deep_gemm` MoE paths for this model that could cost ~1.3 gsm8k points end to end? Things we could not test from outside: decode steps under CUDA-graph replay (padding / `num_token_non_padded` handling in the in-kernel routing), and anything that only shows on real-text activations.

### Reproduction

sglang `0.0.0.dev1+g00143e9c2` (`lmsysorg/sglang:dev` pulled 2026-09-11 01:04 UTC), FlashInfer 0.6.18, 4x GB200 (one NVL72 tray), `zai-org/GLM-5.3-Flash` revision `eb9eb208eb0d988989d07a6a12d0fdeb5f52574a`, lm-evaluation-harness 0.4.13:
```
python -m sglang.launch_server --model-path zai-org/GLM-5.3-Flash --revision eb9eb208eb0d988989d07a6a12d0fdeb5f52574a \
  --served-model-name model --host 127.0.0.1 --port 8000 \
  --tp-size 4 --ep-size 4 --dsa-prefill-backend trtllm --dsa-decode-backend trtllm --kv-cache-dtype fp8_e4m3 \
  --moe-runner-backend flashinfer_trtllm --context-length 16384 --disable-radix-cache
lm_eval --model local-completions \
  --model_args model=model,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=32,max_retries=3,tokenized_requests=False,tokenizer_backend=None \
  --tasks gsm8k --num_fewshot 5 --batch_size 1 --gen_kwargs temperature=0,max_gen_toks=512 --seed 1234 --log_samples --output_path out
```
Replace `--moe-runner-backend flashinfer_trtllm` with `deep_gemm` for the comparison run. Scores above are `exact_match,flexible-extract`. Same-tray throughput cost of deep_gemm on 8k-in/1k-out random prompts: 0.87× (c=1) to 0.98× (c=128) of the trtllm runner.

**Comment by mmangkad** [comment]
@NolenLiang I can reproduce on gsm8k on 4x GB300, every `flashinfer_trtllm` run under every `deep_gemm` run:

| backend | runs | mean | values |
|---|---|---|---|
| `triton` | 1 | 0.9393 | 0.9393 |
| `deep_gemm` | 3 | 0.9335 | 0.9333, 0.9348, 0.9325 |
| `flashinfer_trtllm` | 3 | 0.9260 | 0.9227, 0.9280, 0.9272 |

On a much harder benchmark though, the gap disappears and even flips. GPQA Diamond (198 x 8 repeats, temp 1.0 / top_p 0.95, thinking on), `flashinfer_trtllm` is slightly *ahead*:

| backend | pass@1 | majority@8 | pass@8 |
|---|---|---|---|
| `flashinfer_trtllm` | 0.9154 | 0.9293 | 0.9596 |
| `deep_gemm` | 0.9072 | 0.9091 | 0.9646 |

So it doesn't look like a general regression. I think it's something the shared gsm8k 5-shot prefix triggers.
